zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ServerRegister.java

```diff
@@ -0,0 +1,98 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.zookeeper;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooKeeper;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.zookeeper.NodeValueCodec;
+import com.linecorp.armeria.common.zookeeper.ZKConnector;
+import com.linecorp.armeria.common.zookeeper.ZKListener;
+
+/**
+ * A Server connection maintains the underlying connection and hearing notice from a ZooKeeper cluster.
+ */
+public class ServerRegister {
+    private final ZKConnector zkConnector;
+
+    /**
+     * Create a server register.
+     * @param zkConnectionStr ZooKeeper connection string
+     * @param zNodePath       ZooKeeper node path(under which this server will registered)
+     * @param sessionTimeout  session timeout
+     * @param endpoint        register information
+     * @param nodeValueCodec  nodeValueCodec used
+     */
+    public ServerRegister(String zkConnectionStr, String zNodePath, int sessionTimeout, Endpoint endpoint,
+                          NodeValueCodec nodeValueCodec) {
+        requireNonNull(nodeValueCodec);
+        requireNonNull(endpoint);
+        zkConnector = new ZKConnector(zkConnectionStr, zNodePath, sessionTimeout, new ZKListener() {
+            @Override
+            public void nodeChildChange(Map<String, String> newChildrenValue) {
+            }
+
+            @Override
+            public void nodeValueChange(String newValue) {
+            }
+
+            @Override
+            public void connected() {
+                zkConnector.createChild(endpoint.host() + '_' + endpoint.port(),
+                                        nodeValueCodec.encode(endpoint));
+            }
+        });
+        zkConnector.connect();
+    }
+
+    /**
+     * Create a server register.
+     * @param zkConnectionStr ZooKeeper connection string
+     * @param zNodePath       ZooKeeper node path(under which this server will registered)
+     * @param sessionTimeout  session timeout
+     * @param endpoint        register information
+     */
+    public ServerRegister(String zkConnectionStr, String zNodePath, int sessionTimeout, Endpoint endpoint) {
+        this(zkConnectionStr, zNodePath, sessionTimeout, endpoint, NodeValueCodec.DEFAULT);
+    }
+
+    public void close(boolean active) {
+        zkConnector.close(active);
+    }
+
+    @VisibleForTesting
+    public void enableStateRecording() {
+        zkConnector.enableStateRecording();
+    }
+
+    @VisibleForTesting
+    public ZooKeeper underlyingClient() {
+        return zkConnector.underlyingClient();
+    }
+
+    @VisibleForTesting
+    public BlockingQueue<KeeperState> stateQueue() {
```

Comment:
If these methods are visible only for testing, these methods could be package-local?